website/server/libs/taskManager.js

```diff
@@ -281,3 +291,172 @@ export function moveTask (order, taskId, to) {
+  if (task.type === 'todo') {
+    if (!wasCompleted && task.completed) {
+      // @TODO: mongoose's push and pull should be atomic and help with
+      // our concurrency issues. If not, we need to use this update $pull and $push
+      taskOrderPromise = user.update({
+        $pull: { 'tasksOrder.todos': task._id },
+      }).exec();
+      // user.tasksOrder.todos.pull(task._id);
+    } else if (wasCompleted && !task.completed && user.tasksOrder.todos.indexOf(task._id) === -1) {
+      taskOrderPromise = user.update({
+        $push: { 'tasksOrder.todos': task._id },
+      }).exec();
+      // user.tasksOrder.todos.push(task._id);
+    }
+  }
+
+  setNextDue(task, user);
+
+  if (task.group && task.group.taskId) {
+    await handleSharedCompletion(task);
+  }
+
+  taskScoredWebhook.send(user, {
+    task,
+    direction,
+    delta,
+    user,
+  });
+
+  // Track when new users (first 7 days) score tasks
+  if (moment().diff(user.auth.timestamps.created, 'days') < 7) {
+    res.analytics.track('task score', {
+      uuid: user._id,
+      hitType: 'event',
+      category: 'behavior',
+      taskType: task.type,
+      direction,
+      headers: req.headers,
+    });
+  }
+  return {user, task, delta, direction, taskOrderPromise};
+}
+
+async function handleChallengeTask (task, delta, direction, resolve, reject) {
+  if (task.challenge && task.challenge.id && task.challenge.taskId && !task.challenge.broken && task.type !== 'reward') {
+    // Wrapping everything in a try/catch block because if an error occurs using `await` it MUST NOT bubble up because the request has already been handled
+    try {
+      const chalTask = await Tasks.Task.findOne({
+        _id: task.challenge.taskId,
+      }).exec();
+
+      if (!chalTask) return;
+
+      await chalTask.scoreChallengeTask(delta, direction);
+    } catch (e) {
+      logger.error(e);
+      reject();
+    }
+  }
+  resolve();
+}
+
+export async function scoreTasks (user, taskScorings, req, res) {
+  let taskIds = taskScorings.map(taskScoring => taskScoring.id).filter(id => id !== undefined);
+  if (taskIds.length === 0) throw new BadRequest();
+  let tasks = {};
+  (await Tasks.Task.findMultipleByIdOrAlias(taskIds, user._id)).forEach(task => {
+    tasks[task._id] = task;
+    if (task.alias !== undefined && task.alias.length > 0) {
+      tasks[task.alias] = task;
+    }
+  });
+  if (tasks.length === 0) throw new NotFound(res.t('taskNotFound'));
+  let scorePromises = [];
+  taskScorings.forEach(taskScoring => {
+    if (tasks[taskScoring.id] !== undefined) {
+      scorePromises.push(scoreTask(user, tasks[taskScoring.id], taskScoring.direction, req, res));
+    }
+  });
+  let returnDatas = await Promise.all(scorePromises);
+  if (returnDatas.length === 0) throw new BadRequest();
+  let savePromises = [returnDatas[returnDatas.length - 1].user.save()];
+  returnDatas.forEach(returnData => {
+    if (returnData.taskOrderPromise) savePromises.push(returnData.taskOrderPromise);
```

Comment:
can we combine all the taskOrderPromise in one single update? We risk conflicts by moving multiple things in parallell in separate requests

Reply:
Do you have a suggestion for how to handle that? since it's only pushing and pulling todos, I could have the `returnData` contain if it should be pushed/pulled and then iterate over that to create one user update operation? Does that sound good?

Reply:
Hmm if we want to keep things simple I would probably try to simplify the entire thing by simply creating one API endpoint that accepts multiple requests' data and for each one simulate the scoring of a single task, save the result and then all the results are returned to the client in an array

Reply:
That would make the number of save requests even higher than this version. Since then it would be 2-3 x numberOfTasks (depending on if it's a todo or not and the task has to be moved). Right now my PR saves 1-2 x numberOfTasks + 1. With my proposed change it should be numberOfTasks + 2. (one for the user and one for the task moving)